### DISCUSSION TITLE
Work with NextJS

### DISCUSSION THREAD

**Discussion by qlereboursBS** [OptimalBits/bull#2618] [Q&A]
Hello,

I know there was an issue with this subject, but mine seem to be different, so, I'm creating a discussion about making Bull working with NextJS's API.

## Context 

I was able to make it work once with this code:

First, it needs a specific webpack config in next.config.js to copy lua files:
```
webpack: (config) => ({
    ...config,
    plugins: [
      ...(config.plugins ? config.plugins : []),
      new CopyWebpackPlugin({
        patterns: [
          {
            from: 'node_modules/bull/lib/commands',
            to: 'app/api/translate',
          },
        ],
      }),
    ],
  }),
```

Then, I created a file to POC the queue, in `src/app/api/translator/route.ts`, containing:
 - the queue creation
 - processing
 - a route adding a job

```
// this is for POC purpose
const sleep = (time: number, func: () => void) =>
  new Promise((res) =>
    setTimeout(() => {
      func();
      res(null);
    }, time)
  );

export const translationQueue = new Queue('translation queue', {
  redis: { port: 6379, host: '127.0.0.1' },
});

translationQueue.process(async (job, done) => {
  // job.data contains the custom data passed when the job was created
  // job.id contains id of this job.
  console.log('IN JOB', job.data, job.id);
  await sleep(10000, () => job.progress(42));

  console.log('success');
  return { success: true };
});

export const GET = async (req: Request) => {
  console.debug('In /translate');
  const obj = { age: 12, userId: 1, blabla: { test: '232323' }}
  translationQueue.add(obj);
  console.debug('job added, returning');
  return NextResponse.json({});
};
```

And I ran a redis with  `docker run -p 6379:6379 -it redis/redis-stack-server:latest`

## Issue

When I call the endpoint that add a job, it works only the first time:
Here are the logs produced:
```
In /translate
job added, returning
IN JOB { age: 12, userId: 1, blabla: { test: '232323' } } 1
success
In /translate
job added, returning
IN JOB { age: 12, userId: 1, blabla: { test: '232323' } } 2
```

I will test with something else than NextJS to see if it's a misconfiguration or if it's NextJS

**Comment by manast** [comment]
As stated in other issues and in the documentation, do not mix "done" callback and return promises in the processor function. Btw, you could also upgrade to BullMQ which only supports async processors to avoid this error happen in the future.

**Reply by qlereboursBS** [reply]
Ok thank you very much for your help. I upgraded to BullMQ and everything's working well, sorry for the beginner issue
